internal/provider/kubernetes/controller.go

```diff
@@ -782,6 +769,56 @@ func (r *gatewayAPIReconciler) findReferenceGrant(ctx context.Context, from, to
 	return nil, nil
 }
 
+// processFinalizers encapsulates logic for managing finalizers on GatewayClass and EnvoyProxy objects.
+func (r *gatewayAPIReconciler) processFinalizers(ctx context.Context, managedGC *gwapiv1.GatewayClass, resourceTree *gatewayapi.Resources) error {
+	// Add finalizer to EnvoyProxy if it is referenced by the GatewayClass.
+	if classRefsEnvoyProxy(managedGC, resourceTree.EnvoyProxyForGatewayClass) && resourceTree.EnvoyProxyForGatewayClass.DeletionTimestamp.IsZero() && managedGC.DeletionTimestamp.IsZero() {
+		if err := r.addFinalizer(ctx, resourceTree.EnvoyProxyForGatewayClass); err != nil {
+			return fmt.Errorf("failed adding finalizer to Envoy Proxy %s: %w", resourceTree.EnvoyProxyForGatewayClass.Name, err)
+		}
+		if err := r.addFinalizer(ctx, managedGC); err != nil {
```

Comment:
Do we need to add a finalizer to the GatewayClass here?

Reply:
I want to add it to GatewayClass, when we reference EnvoyProxy and no Gateways are attached.
This is needed for situation, when GatewayClass is deleted before EnvoyProxy

Reply:
Sorry, I may miss something here. Would it be a problem if a GatewayClass is deleted when its referencing EnvoyProxy exists?

Reply:
No, it is only checking if `EnvoyProxy` exists for the given `GatewayClass`.
As for now, it is working correctly for both scenarios, so `GatewayClass` deletion will hold until Gateways are cleaned up and then finalizers are removed from `GatewayClass` and `EnvoyProxy`.